chapters/lexicalstructure.tex

```diff
@@ -247,6 +247,24 @@ \subsection{Strings}\label{strings}
 \end{lstlisting}
 \end{nonnormative}
 
+\subsection{Units of Literal Constants}\label{units-literal-constants}
+
+The following rules apply:
+- The result of `x + L`, `L + x`, `x - L`, `L - x`, `x*L`, `L*x`, `x/L`, where `x` is a variable with non-empty \lstinline!unit! attribute string `<s>` and `L` is an expression containing only literal constants, shall have the same \lstinline!unit! attribute string of `x`.
```

Comment:
Just to reiterate the previous discussion. Including, `x*L` and `L*x` here means that the following is valid, as @henrikt-ma  pointed out:
```
Real x(unit = "m") = 1;
Real y(unit = "m2") = x * 10^2; /* OK, due to wildcard effect of 10^2. */
```

Reply:
AIUI, this is accounted for by the definition that
> `L` is an _expression_ containing only literal constants

`10^2` is an expression containing only literal constants (at least, if you consider the operators in an expression not to count -- maybe better would be "`L` is an _expression_ operating only on literal constants).

Reply:
To avoid having to keep in mind/memory the whole expression tree below a given leaf to be able to make the right decision (which could be complicated/expensive), maybe we can specify that the unitness of an expression that operates purely on literal constants is equivalent to a literal constant (with some edge cases, surely).

Initial proposal for an additional rule:
```
- The result of mathematical operations *(precise scope TBD)* purely on literal constants does not have a defined, non-empty \lstinline!unit! string attribute. (Non-normative addition: they behave just like literal constants themselves with respect to unit handling)
```

WIP scope of "mathematical operations" (help me refine, pls)
* Elementary mathematical functions (3.7.3)
* Arithmetic operators (3.4)

Reply:
> Thoughts?

The two proposals differ substantially in how far they take unit checkin, but with the recent development of this one I don't really see a reason why the two proposals couldn't both do it in the "standard" recursive bottom-up fashion.
I'm ready to port any refinements in this PR over what is currently in #3257, so that get get this detail out of the way when comparing the proposals.

Reply:
> AIUI, this is accounted for by the definition that
> 
> > `L` is an _expression_ containing only literal constants
> 
> `10^2` is an expression containing only literal constants (at least, if you consider the operators in an expression not to count -- maybe better would be "`L` is an _expression_ **operating only on** literal constants).

I don't understand how what you are saying relates to my comment.

My interpretation of the current wording is that `10^2` would receive the same unit as `x`, in the example above that would be `m` which would make it valid. I would prefer if our specs could reject the given example.

Reply:
> AIUI, this is accounted for by the definition that
> 
> > `L` is an _expression_ containing only literal constants

Yes. This is not clearly explained in the text committed on this PR branch.

> `10^2` is an expression containing only literal constants (at least, if you consider the operators in an expression not to count -- maybe better would be "`L` is an _expression_ **operating only on** literal constants).

I'm not sure if the Modelica specification itself has rigorous definitions of terms such as _expression operating on something_. I only found one [here](https://specification.modelica.org/maint/3.5/arrays.html), the problem is it clearly refers to expressions where operators are used. As I wrote in a previous comment, I see no reason why we shouldn't also include function calls (both proper functions and function-like operators alike). So, maybe "containing only literal constants" is clear enough for that?

Reply:
The problem with functions is that many functions have units in the interface, e.g. `specificEnthalphy_pT(1e5, 293.15)` which clearly shouldn't be treated as if it were only literal constants.

The interesting part with `sin` is that `Modelica.Math.sin` similarly has unit in the interface - and it would be nice if both were treated the same.

Reply:
> To avoid having to keep in mind/memory the whole expression tree below a given leaf to be able to make the right decision (which could be complicated/expensive),

The proposed text does not imply that. This is an implementation detail. As I understand, there is no need to keep the whole tree in memory, you just need to carry the information about sub-expressions only involving literals with undefined units as you process the expression tree.

> - The result of mathematical operations *(precise scope TBD)* where all operands have either undefined or empty  \lstinline!unit! string attribute has an empty \lstinline!unit! string attribute. 

Wait: undefined unit (i.e. literal) is not the same as empty unit string (e.g. Real variable or parameter). We are only concerned about literals with undefined unit in this proposal.

> Then, we can replace "L is an expression containing only literal constants" with "L has undefined or empty unit string", 

Only undefined. If it is empty, then it is correct to consider it as wildcard. Anyway, I wouldn't go down this rabbit hole. If you want to do something more than what I proposed here, then it's best to work on the other proposal, which is a lot more ambitious.

More in general, I would try not to overkill over such a tiny issue 😅 .

Reply:
> My interpretation of the current wording is that `10^2` would receive the same unit as `x`, in the example above that would be `m` which would make it valid. I would prefer if our specs could reject the given example.

The current wording implies that `x*10^2` and `10^2*x` get the same unit as `x`. It's the result of the multiplication or addition that retains the unit, not the literal constant taking any unit.

Reply:
> > My interpretation of the current wording is that `10^2` would receive the same unit as `x`, in the example above that would be `m` which would make it valid. I would prefer if our specs could reject the given example.
> 
> The current wording implies that `x*10^2` and `10^2*x` get the same unit as `x`. It's the result of the multiplication or addition that retains the unit, not the literal constant taking any unit.

I see

Reply:
> We would however be rejecting the following

See below.

> Real x1(unit = "m") = 1;

The RHS has undefined unit, so there is no reason why this should be rejected. No dimensional inconsistency can be detected between a length and a quantity with undefined unit. Do I miss something?

> Real x2(unit = "s") = 2

Ditto.

> Real y(unit = "m/s") = (x1/x2) * 10^2; /* Rejected because 10^2 will receive "m/s". */

`x1/x2` has unit "1". Hence `x1/x2*10^2` will retain the same unit, causing a rejection because you have "m/s" on the LHS and "1" on the RHS. This is consistent with my requirements: since there are quantities with non-empty unit strings x1 and x2, 10^2 should behave as a dimensionless constant, not as a dimensional wildcard.

Thinking about it, we should probably improve the wording of the basic rules from "where `x` is a variable with non-empty \lstinline!unit! attribute string `<s>`" to "where `x` is an expression with non-empty \lstinline!unit! attribute string `<s>`". I'll update the PR accordingly.

Reply:
Done.

Reply:
> The proposed text does not imply that. This is an implementation detail. As I understand, there is no need to keep the whole tree in memory, you just need to carry the information about sub-expressions only involving literals with undefined units as you process the expression tree.

Ah, interesting context! I failed to consider that implementations could separately/on their own track if subexpressions only involved literals even _without_ support for this in the spec. Hence my intention to specify that.

In that case, I think we can forget about the concern about propagating "literal-onlyness" upwards.

Reply:
> `x1/x2` has unit "1".

I think you got mixed up -- `x1/x2` has units "m" / "s".

Reply:
OK, you are right, then it will be accepted, as far as I understand - multiplying an expression with unit "m/s" by a unit-less literal keeps the same unit.

Reply:
AFAICT, the sole remaining issue to resolve in this conversation thread is how to deal with calls of functions with units in their interface.

Francesco:
> As I wrote in a previous comment, I see no reason why we shouldn't also include function calls (both proper functions and function-like operators alike). So, maybe "containing only literal constants" is clear enough for that?

Hans:
> The problem with functions is that many functions have units in the interface, e.g. `specificEnthalphy_pT(1e5, 293.15)` which clearly shouldn't be treated as if it were only literal constants.
> 
> The interesting part with `sin` is that `Modelica.Math.sin` similarly has unit in the interface - and it would be nice if both were treated the same.

Reply:
For proper functions, the rules that we have specified for handling the units of expressions containing literals should apply. Say that specificEnthalpy_pT(p,T) wants an input p with unit "Pa" and an input T with unit "K", and that we have defined variables p and T with such units, and h with unit "J/kg".
- specificEnthalpy_pT(p, T) is accepted, because the input expressions have a well-defined unit, which is the right one
- specificEnthalpy_pT(p, h) is rejected, because the second input has a well-defined unit, which is different from what is expected
- specificEnthalpy_pT(2*p, T) is accepted, because 2*p has the same unit of p, i.e. "Pa"
- specificEnthalpy_pT(p, T+2) is accepted, because T+2 has the same unit of T, i.e. "K"
- specificEnthalpy_pT(p, h/1000) is rejected, because h/1000 keeps the unit of h, i.e. "J/kg"
- specificEnthalpy_pT(1e5, 293.15) is accepted, because the inputs have no units information, so you can't say if they are dimensionally inconsistent
- specificEnthalpy_pT(1e5, 293.15 + time*20) is rejected, because the second input has unit "s" instead of "K"
- h + specificEnthalpy_pT(p,T) is accepted because it sums two quantities with unit "J/kg"
- h + specificEnthalpy_pT(1e5, 293.15) is accepted because it sums two quantities with unit "J/kg"
- T + specificEnthalpy_pT(p,T) is rejected because is sums a quantity with unit "K" and a quantity with unit "J/kg".
- 1e3 + specificEnthalpy_pT(p,T) is accepted because L + x has the same unit of x, i.e., "J/kg"

From this point of view, I don't think we need to say anything specific about regular functions, what to do with them when doing dimensional consistency checking is obvious: if you have defined unit information, use it for checking, otherwise don't do anything.

The reason built-in functions are mentioned explicitly is that they _don't have unit information defined anywhere_. With the current formulation of the rules, maybe the most appropriate place to say that is not here, but rather [section 3.7.3](https://specification.modelica.org/master/operators-and-expressions.html#built-in-mathematical-functions-and-external-built-in-functions). What do you think?

Reply:
NB: It's not sufficient to restrict our rule application to `arithmetic-expression` instead of "expression", because according to the grammar (Spec A.2.7), the function call syntax is part of the `primary` token, so `specificEnthalpy_pT(1e5, 293.15)` is an `arithmetic-expression` if I read the grammar correctly.

Reply:
@bilderbuchi good point. I assume we all agree that the expected results I posted in my [previous comment](https://github.com/modelica/ModelicaSpecification/pull/3266#discussion_r1022592658) are OK. It shows that "expression containing only literal constants" is a bit sloppy and simply not good enough.

The important point here is that expressions are built using operators, function-like operators and regular function. For some of them, how units must be handled is pretty obvious, e.g. in a + b, a*b, a/b, sqrt(b), smooth(0, expr), fill(expr, n, m) and, most importantly, _the unit of the result depends on the units of the operands / function inputs_. All these functions should be considered in the same way as operators such as `+` or `*`.

Then, there are functions such as specificEnthalpy_pT(), where instead the unit of the output is actually declared in the function itself, so _it does not (and should not)_ depend on the unit of the inputs. These need to be handled in a completely differnt way. In fact, these functions act as watersheds for unit checking: you should separately check that their inputs are fine (i.e., if they have a defined, non-empty unit, it has to be the right one), and then their output can be considered as a variable with the unit of their output.

We need a way to tell these two categories apart and properly convey this context. At the moment, I'm not really sure how to do that in a concise way. Maybe you can make a proposal for that.

Reply:
Agreed on all points.

> Maybe you can make a proposal for that.

My best shot so far has been the approach proposed [above](https://github.com/modelica/ModelicaSpecification/pull/3266#discussion_r1018756734), i.e. to define the unitness of the result of purely-literal (or empty-unit) operations, and use that newly gained definition in the present operation rules for mixed-unitness operands. 
This would liberate us from having to distinguish between the two categories you mention, as now for all operands we have well-defined unitness, and unit checking can follow the expression tree.

I'd be happy to iterate more on this approach, but you considered this out of scope (as it's not concerned only with literals anymore), and too ambitious, so I did not follow up further. Should I pick this up again and give it another go?

Reply:
@bilderbuchi yes, that was the idea. ""L is an expression containing only literal constants" is really weak.

However, I'm not sure what you proposed [above](https://github.com/modelica/ModelicaSpecification/pull/3266#discussion_r1018756734) is OK. You wrote "L has undefined or empty unit string", but allowing empty strings promotes them to the same level of "dimensionless literals" (which are not explicitly defined, but you get the idea). This may lead to undesirable side effects, e.g.:
```
Real L = 1;
Real A(unit="m2") = 2;
Real V(unit="m3") = A*L;
```
here `L` has empty unit string, so `A*L` will be considered to have unit "m2", and then the third binding equation would be rejected.

The goal of this PR is just to avoid the unit wildcard effect of literal units (that have undefined unit string), without introducing non-backwards compatible changes to unit checking. If one mixes _variables_ with and without units, the tool should just give up dimensional consistency checking. As I understand, your [last proposal](https://github.com/modelica/ModelicaSpecification/pull/3266#discussion_r1018756734) doesn't really achieve that.

Reply:
> allowing empty strings promotes them to the same level of "dimensionless literals"

Indeed. My intent was to "unify" behaviour for "unitless" (undefined or empty) objects. Maybe I had a wrong picture of what "proper" behaviour is.

_Please disregard the question if it fits in this proposal or not for a moment_. 
If I understand your remark on undesirable side-effects correctly, you want *variables* with empty unit string to *retain* a wildcard effect, so the equation for `V` should be accepted althought it has m3 on the left and m2*number on the right. Correct?

My view was that having a `Real L = 2` with empty unit string is *effectively* just a label/_reference_ for a string literal (for which we want to avoid wildcard behaviour). So, I thought that empty-unit variables and undefined-unit literals should effectively behave the same, which would mean that the equation for `V` should be rejected like the one for `V2`.
```
Real L = 1;
Real A(unit="m2") = 2;
Real V(unit="m3") = A*L; // you want this accepted (L is a wildcard)?
Real V2(unit="m3") = A*2; // should be rejected (m3 != m2)
```

I'm fully in the clear about your desire to absolutely minimize the scope of this proposal to increase its chances of being accepted, and I agree with you! 
However, we now seem to have hit a point where we can't find a watertight formulation for how to specify our intent. This _alone_ prompted me to pick up the previous proposal, to see if we can salvage something from it. Maybe a small extension of the scope will lead to a better-expressible specification, and reach agreement or at least discussion?

Let me know your thoughts, and I'll iterate more on it.

Reply:
> Indeed. My intent was to "unify" behaviour for "unitless" (undefined or empty) objects. Maybe I had a wrong picture of what "proper" behaviour is.
> 
> _Please disregard the question if it fits in this proposal or not for a moment_.

Motion of order: if you want to discuss another proposal, please start another PR 😃 . This discussion is way too long already.

The goal of this PR is to avoid the wildcard effect of literal constants, avoiding any kind of semantic change that causes existing code (which is not plain wrong or highly questionable, of course) to be rejected by dimensional checking. So, to me the proper behaviour is that if `A` has unit "m2" and L has unit "", `2*A` must have unit "m2", while it won't be possible to say anything about the unit of `A*L`, because L has an empty unit string.

People have been writing code with non fully specified units for ages. Sometimes for good reasons, e.g. when we wrote fluid models with weird partial derivatives, we didn't bother to specify their units in detail, because that was boring and we may have actually made mistakes while doing that, creating bogus dimensional checking errors. So, we left them with unit = "". It is not compulsory to use units in Modelica, by any means, so by no means variables with unit = "" should be interpreted as if they were dimensionless. This has never been the intention.

> If I understand your remark on undesirable side-effects correctly, you want _variables_ with empty unit string to _retain_ a wildcard effect

Yes.

> so the equation for `V` should be accepted although it has m3 on the left and m2*number on the right. Correct?

Yes. Otherwise we break backwards compatibility badly.
 
> My view was that having a `Real L = 2` with empty unit string is _effectively_ just a label/_reference_ for a string literal

No, it's not. This is the whole point of this proposal: a literal unit used in an expression _is not the same_ as a variable with empty unit attribute, which is de-facto how it is interpreted by today's tools. Regardless of how it has been defined/computed.

> So, I thought that empty-unit variables and undefined-unit literals should effectively behave the same, which would mean that the equation for `V` should be rejected like the one for `V2`.

No. If we reject the equation for `V`, I bet there will be hundreds of cases in the MSL that will be rejected, beyond the Spice library which is a bit special because it's a straight porting from FORTRAN code. That's not what I want with this PR. And I guess @HansOlsson will be even more opposed to it 😅.

> I'm fully in the clear about your desire to absolutely minimize the scope of this proposal to increase its chances of being accepted, and I agree with you! However, we now seem to have hit an impasse, where we cannot find a watertight formulation for how to specify our intent.

I am sure we can find a better formulation for that with a bit of patience, _without changing the scope of the proposal_. The case of expressions involving literal constants and mathematical operators is pretty clear. We just need to define how to handle function calls and function-like operators. The idea is explained in a bit loose terms in my [previous comment](https://github.com/modelica/ModelicaSpecification/pull/3266#discussion_r1023048775). We just need to make it shorter and as formal as possible.

If you want something stronger, that would a different PR, and maybe you should consider the alternative proposal by @henrikt-ma first 😃.

Reply:
😄

Reply:
I have an updated proposal, see the separate thread https://github.com/modelica/ModelicaSpecification/pull/3266/files#r1049500278 (for a better viewing experience, this one is huge).